modules/catalog/src/main/java/org/apache/ignite/internal/catalog/commands/RemoveIndexCommand.java

```diff
@@ -0,0 +1,97 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.ignite.internal.catalog.commands;
+
+import static org.apache.ignite.internal.catalog.commands.CatalogUtils.indexOrThrow;
+import static org.apache.ignite.internal.catalog.descriptors.CatalogIndexStatus.STOPPING;
+
+import java.util.List;
+import org.apache.ignite.internal.catalog.Catalog;
+import org.apache.ignite.internal.catalog.CatalogCommand;
+import org.apache.ignite.internal.catalog.CatalogValidationException;
+import org.apache.ignite.internal.catalog.descriptors.CatalogIndexDescriptor;
+import org.apache.ignite.internal.catalog.descriptors.CatalogIndexStatus;
+import org.apache.ignite.internal.catalog.storage.RemoveIndexEntry;
+import org.apache.ignite.internal.catalog.storage.UpdateEntry;
+
+/**
+ * A command that removes index with specified ID from the Catalog. Only makes sense for an index in the {@link CatalogIndexStatus#STOPPING}
+ * state (so it's about removing a dropped index from the Catalog when we don't need it anymore).
+ *
+ * <p>This is always invoked as a reaction to an internal trigger, not directly by the end user.
+ *
+ * <p>For dropping an index, please refer to {@link DropIndexCommand}.
+ *
+ * @see DropIndexCommand
+ */
+public class RemoveIndexCommand implements CatalogCommand {
+    /** Returns builder to create a command to remove index with specified name. */
+    public static RemoveIndexCommandBuilder builder() {
+        return new Builder();
+    }
+
+    private final int indexId;
+
+    private final String schemaName;
```

Comment:
I thought the same way, but there is a tricky case when an index is removed when its table has already been removed from the Catalog. This can happen in the following scenario:

1. A table exists with an index that is in the AVAILABLE state
2. DROP INDEX happens, so it moves to the STOPPING state
3. DROP TABLE happens, so the table is removed from the Catalog
4. After some time (after waiting out all RW txs started before the index drop), we try to remove the index from the Catalog

To remove the index, we need to find its schema. We cannot find the schema using the table, because the table does not exist in the Catalog anymore. This is why I added `schemaName` here.

On the other hand, `STOPPING` state on an index is only needed to let old RW transactions still write to the index. After the table is dropped, all such writes are immediately forbidden. So, on table drop, we can forcibly remove any STOPPING index from the Catalog (and do the same with AVAILABLE indexes of the table, skipping the STOPPING state). If we do this, no index would be able to exist in the Catalog for a table that is removed from the Catalog, so the problem would go away, and the `schemaName` would become redundant.

I'll try to implement this.

Reply:
We leave it as is, we discussed it in person